test_runner_lib/include/test_runner/TestOutputSpec.h

```diff
@@ -0,0 +1,156 @@
+  std::vector<RegisterPrecondition> register_preconditions;
+  std::vector<RegisterPrecondition> register_postconditions;
+  std::vector<MemoryModifier> initial_memory_conditions;
+  std::vector<MemoryModifier> expected_memory_conditions;
+  std::unordered_map<std::string, std::function<std::any(State &)>>
+      reg_to_accessor;
+
+  template <typename T>
+  void ApplyCondition(State &state, std::string reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      std::any_cast<std::reference_wrapper<T>>(accessor->second(state)).get() =
+          value;
+    }
+  }
+
+  template <typename T>
+  void CheckCondition(State &state, std::string reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      auto actual =
+          std::any_cast<std::reference_wrapper<T>>(accessor->second(state));
+      LOG(INFO) << "Reg: " << reg << " Actual: " << std::hex
+                << static_cast<uint64_t>(actual.get())
+                << " Expected: " << std::hex << static_cast<uint64_t>(value);
+      CHECK_EQ(actual, value);
+    }
+  }
+
+ public:
+  template <typename T>
+  void AddPrecWrite(uint64_t addr, T value) {
+    this->initial_memory_conditions.push_back(
+        [addr, value](MemoryHandler &mem_hand) {
+          mem_hand.WriteMemory(addr, value);
+        });
+  }
+
+  template <typename T>
+  void AddPostRead(uint64_t addr, T value) {
+    this->expected_memory_conditions.push_back(
+        [addr, value](MemoryHandler &mem_hand) {
+          LOG(INFO) << "Mem: " << std::hex << addr << " Actual: " << std::hex
+                    << mem_hand.ReadMemory<T>(addr) << " Expected: " << std::hex
+                    << value;
+          CHECK_EQ(mem_hand.ReadMemory<T>(addr), value);
+        });
+  }
+
+  const std::vector<MemoryModifier> &GetMemoryPrecs() const {
+    return this->initial_memory_conditions;
+  }
+
+  const std::vector<MemoryModifier> &GetMemoryPosts() const {
+    return this->expected_memory_conditions;
+  }
+
+  TestOutputSpec(
+      uint64_t disas_addr, std::string target_bytes,
+      remill::Instruction::Category expected_category,
+      std::vector<RegisterPrecondition> register_preconditions,
+      std::vector<RegisterPrecondition> register_postconditions,
+      std::unordered_map<std::string, std::function<std::any(State &)>>
+          reg_to_accessor)
+      : addr(disas_addr),
+        target_bytes(target_bytes),
+        expected_category(expected_category),
+        register_preconditions(std::move(register_preconditions)),
+        register_postconditions(std::move(register_postconditions)),
+        reg_to_accessor(reg_to_accessor) {}
+
+
+  void SetupTestPreconditions(State &state) const {
+    for (auto prec : this->register_preconditions) {
+      std::visit(
+          [this, &state, prec](auto &&arg) {
+            using T = std::decay_t<decltype(arg)>;
```

Comment:
Instead of this newer C++ (remill should support it, but not 100%sure) gives you
```
[this, &state, prec]< typename T >( T && arg ) { ...
```